### PR TITLE
feat: Python compatibility

### LINKED ISSUE #2248: Python compatibility
Detect (a) version of Python in use, and (b) of any updates are incompatible.

### PR DESCRIPTION
**Pull request by kayoub5**
Closes #2248

### CODE REVIEW

**Comment by rarkins** [comment]
Questions:

1. Is there a way this can be dynamically detected for python?
2. Are you sure about making it an object, rather than a plain string field that can then be configured *inside* an object, e.g. like below?

```
{
  "python": {
    "compatibilityRestriction": "2.7"
  },
  "php": {
    "compatibilityRestriction": "5.1"
  }
}

**Comment by kayoub5** [comment]
@rarkins 
> Is there a way this can be dynamically detected for python?
For `pipenv` yes, for `requirements.txt`, I don't think so.
> Are you sure about making it an object, rather than a plain string field that can then be configured inside an object, e.g. like below?
for now we only support the python version, but more restrictions could be included, such as platform (windows/ linux), distribution (cpython, pypy, cython, ironpython...).

File: lib/datasource/pypi.js [hunk 1]
```diff
@@ -69,9 +87,10 @@ async function getPkgReleases(purl, config = {}) {
       dependency.repositoryUrl || manualRepositories[depName.toLowerCase()];
     dependency.releases = [];
     if (dep.releases) {
-      const versions = Object.keys(dep.releases)
-        .filter(isVersion)
-        .sort(sortVersions);
+      const versions = compatibleVersions(
+        dep.releases,
+        compatibilityRestriction
+      );
```
[the remark this answers is not in the document]
Comment (kayoub5) [reply lib/datasource/pypi.js]: @rarkins caching is not within the scope of the original issue #2248, do you think it's really important to include it here, or could be deferred to its own PR?

[the remark this answers is not in the document]
Comment (rarkins) [reply lib/datasource/pypi.js]: Caching can be deferred, it’s just we’ll need to refactor a little of the new code

**Comment by rarkins** [comment]
Good point to make it generic, although we may want to semantically document the possible keys Eg for json schema. What do you think about renaming it to be something shorter like “compatibility” or “restrictions”?

**Comment by kayoub5** [comment]
@rarkins  done

**Pull request merged**